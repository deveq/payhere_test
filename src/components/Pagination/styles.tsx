import styled from 'styled-components';
import {flexCenter} from 'styles/helper';

interface PageButtonProps {
  selected?: boolean;
  hide?: boolean;
}

export const Container = styled.nav`
    ${flexCenter};
    gap: 4px;
    height: 20px;
`;

export const PageButton = styled.button<PageButtonProps>`
  border: none;
  border-radius: 8px;
  background: ${({theme, selected}) => selected ? theme.primary : 'transparent'};
  color: ${({theme, selected}) => selected ? 'white' : theme.text1};
  visibility: ${({hide}) => hide ? 'hidden' : 'visible'};
  &:hover {
    transform: translateY(-2px);
    cursor: pointer;
  }
  &[disabled] {
    cursor: revert;
    transform: revert;
  }
`;
