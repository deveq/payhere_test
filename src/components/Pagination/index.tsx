import { SetStateAction, Dispatch } from 'react';
import * as Styled from './styles';

export interface PaginationProps {
	total: number;
	page: number;
	setPage: Dispatch<SetStateAction<number>>;
	limit: number;
}

const Pagination = ({ page, setPage, limit, total }: PaginationProps) => {
	const numPages = Math.ceil(total / limit);
	const startPage = Math.floor((page - 1) / 10) * 10 + 1;
	const endPage = Math.min(startPage + 9, numPages);

	const getNumber = () => {
		const arr = [];
		for (let i = startPage; i <= endPage; i++) {
			arr.push(i);
		}
		return arr;
	};

	const moveToPrev = () => {
		setPage((prev) => prev - 1);
	};

	const moveToNext = () => {
		setPage((prev) => prev + 1);
	};

	const moveToTarget = (num: number) => {
		setPage(num);
	};

	return (
		<Styled.Container>
			<Styled.PageButton onClick={moveToPrev} disabled={page === 1}>
				&lt;
			</Styled.PageButton>
			{startPage > 10 && (
				<>
					<Styled.PageButton
						onClick={() => {
							moveToTarget(1);
						}}
					>
						1
					</Styled.PageButton>
					<Styled.PageButton disabled>...</Styled.PageButton>
				</>
			)}
			<>
				{getNumber().map((num) => (
					<Styled.PageButton
						onClick={() => {
							moveToTarget(num);
						}}
						selected={num === page}
						key={`page_${num}`}
					>
						{num}
					</Styled.PageButton>
				))}
			</>
			<>
				{endPage !== numPages && (
					<>
						<Styled.PageButton disabled>...</Styled.PageButton>
						<Styled.PageButton
							onClick={() => {
								moveToTarget(numPages);
							}}
						>
							{numPages}
						</Styled.PageButton>
					</>
				)}
			</>
			<Styled.PageButton onClick={moveToNext} disabled={page === numPages}>
				&gt;
			</Styled.PageButton>
		</Styled.Container>
	);
};

export default Pagination;

/*
total, limit을 받고 10개씩 페이지를 렌더링 한다
numPages가 10 이상일 경우 앞 혹은 뒤에 ...을 렌더링해주는데
제일 앞, 뒤에는 1번과 마지막 페이지를 렌더링 해준다

case1. total: 132, limit: 10, curr: 1
< 1 ~ 10 ... 14 >
case2. total: 132, limit: 10, curr 11
< 1 ... 11 ~ 14 >
case3. total: 632, lmit: 10, curren 33
< 1 ... 31 ~ 40 ... 64 >

1이면 0
11이면 1
21이면 2
const startPage = Math.floor(currentPage / 10) * 10 + 1 부터 시작

startPage
pageCount / 10

만약 currentPage가 pageCount보다 크다면 1페이지를 렌더링하도록

//계산
const pageCount = Math.ceil(total / 10) = 14;

TODO: 버그발견
current page가 20일때 11~20이 아니라 21~30으로 렌더링됨!


 */
