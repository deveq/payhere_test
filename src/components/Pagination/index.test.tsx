
import {screen, fireEvent} from '@testing-library/react';
import {useState} from 'react';
import {renderHook, act} from '@testing-library/react-hooks';
import userEvent from '@testing-library/user-event';
import 'jest-styled-components';
import {renderWithWrapper} from 'tests';
import Pagination from './index';

describe('<Pagination />', () => {
    it('정상적으로 렌더링되었는지 확인한다', () => {
        const {result: {current: [page, setPage]}} = renderHook(() => useState(1));
        const {container} = renderWithWrapper(<Pagination 
            total={132}
            page={page}
            setPage={setPage}
            limit={10}
            onlyArrow={true}
        />);

        const pagination = screen.getByRole('navigation');
        expect(pagination).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    })
    
    it('only Arrow Pagination이 정상렌더링 되는지 확인한다', () => {
        // const {result: {current: [page, setPage]}} = renderHook(() => useState(1));
        const {result} = renderHook(() => useState(1));
        const [page, setPage] = result.current;
        const {container} = renderWithWrapper(<Pagination 
            total={132}
            page={page}
            setPage={setPage}
            limit={10}
            onlyArrow={true}
        />);
    
        const pagination = screen.getByRole('navigation');
        const prevButton = screen.getByText('이전 페이지');
        const nextButton = screen.getByText('다음 페이지');

        expect(pagination.childElementCount).toBe(2);

        expect(prevButton).toBeInTheDocument();
        expect(nextButton).toBeInTheDocument();
    })

    it('page를 11이상으로 변경한다.', () => {
        const {result: {current: [page, setPage]}} = renderHook(() => useState(12));
        
        renderWithWrapper(<Pagination
            limit={10}
            page={page}
            setPage={setPage}
            total={638}
        />)

        const pagination = screen.getByRole('navigation');
        // const buttons = screen.getAllByRole('button');
        const button = screen.getByText(13);
        userEvent.click(button);

        expect(pagination).toBeInTheDocument();
    })
})

// export interface PaginationProps {
// 	onlyArrow?: boolean;
// }

// interface PageButtonProps {
//   selected?: boolean;
//   hide?: boolean;
// }

// export const Container = styled.nav`
// `;

// export const PageButton = styled.button<PageButtonProps>`
//   background: ${({theme, selected}) => selected ? theme.primary : 'transparent'};
//   color: ${({theme, selected}) => selected ? 'white' : theme.text1};
//   visibility: ${({hide}) => hide ? 'hidden' : 'visible'};
//   &:hover {
//     transform: translateY(-2px);
//     cursor: pointer;
//   }
//   &[disabled] {
//     cursor: revert;
//     transform: revert;
//   }
// `;
