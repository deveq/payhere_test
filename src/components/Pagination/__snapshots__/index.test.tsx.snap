// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> 정상적으로 렌더링되었는지 확인한다 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 4px;
  height: 20px;
}

.c1 {
  border: none;
  border-radius: 8px;
  background: transparent;
  color: #212529;
  visibility: hidden;
}

.c1:hover {
  -webkit-transform: translateY(-2px);
  -ms-transform: translateY(-2px);
  transform: translateY(-2px);
  cursor: pointer;
}

.c1[disabled] {
  cursor: revert;
  -webkit-transform: revert;
  -ms-transform: revert;
  transform: revert;
}

.c2 {
  border: none;
  border-radius: 8px;
  background: transparent;
  color: #212529;
  visibility: visible;
}

.c2:hover {
  -webkit-transform: translateY(-2px);
  -ms-transform: translateY(-2px);
  transform: translateY(-2px);
  cursor: pointer;
}

.c2[disabled] {
  cursor: revert;
  -webkit-transform: revert;
  -ms-transform: revert;
  transform: revert;
}

<div>
  <nav
    class="c0"
  >
    <button
      class="c1"
      disabled=""
    >
      이전 페이지
    </button>
    <button
      class="c2"
    >
      다음 페이지
    </button>
  </nav>
</div>
`;
